DOC-215 | Clarify difference between UPDATE/REPLACE syntaxes (#1181)
arangodb/docs: 3.10/aql/operations-update.md

Both variants can optionally end with an `OPTIONS { … }` clause.

`collection` must contain the name of the collection in which the documents should
be updated. `document` must be an object that contains the attributes and values 
to update. Attributes that don't yet exist in the stored document are added to it.

### `UPDATE <document> IN <collection>`

Using the first syntax, the `document` object must have a `_key` attribute with
the document key. The existing document with this key is updated with the
attributes provided by the `document` object.

The following query adds or updates the `name` attribute of the document
identified by the key `my_key` in the `users` collection. The key is passed via
the `_key` attribute alongside other attributes:

```aql
UPDATE { _key: "my_key", name: "Jon" } IN users
```

The following query is invalid because the object does not contain a `_key`

> attribute and thus it is not possible to determine the documents that needs to

Reviewer: ```suggestion
attribute and thus it is not possible to determine the documents to
```

```aql
UPDATE { name: "Jon" } IN users
```

You can combine the `UPDATE` operation with a `FOR` loop to determine the
necessary key attributes, like shown below:

```aql
FOR u IN users
  UPDATE { _key: u._key, name: CONCAT(u.firstName, " ", u.lastName) } IN users
```

Note that the `UPDATE` and `FOR` operations are independent of each other and
`u` does not automatically define a document for the `UPDATE` statement.
Thus, the following query is invalid:

```aql
FOR u IN users
  UPDATE { name: CONCAT(u.firstName, " ", u.lastName) } IN users
```

### `UPDATE <keyExpression> WITH <document> IN <collection>`

Using the second syntax, the document to update is defined by the
`keyExpression`. It can either be a string with the document key, an object
which contains a `_key` attribute with the document key, or an expression that
evaluates to either of these two.

The following query adds or updates the `name` attribute of the document
identified by the key `my_key` in the `users` collection. The key is passed as
a string in the `keyExpression`. The attributes to add or update are passed
separately as the `document` object:

```aql
UPDATE "my_key" WITH { name: "Jon" } IN users
```

> If the `document` object may contain a `_key` attribute but it is ignored.

Reviewer: An extra "If".

> If the `document` object may contain a `_key` attribute but it is ignored.

Reviewer: ```suggestion
The `document` object may contain a `_key` attribute, but it is ignored.
```

  UPDATE u WITH { name: CONCAT(u.firstName, " ", u.lastName) } IN users
```

Dynamic key expressions
-----------------------

An `UPDATE` operation may update arbitrary documents, using either of the two
syntaxes:

```aql
FOR i IN 1..1000
  UPDATE { _key: CONCAT('test', i), name: "Paula" } IN users
```

```aql
FOR i IN 1..1000
  UPDATE CONCAT('test', i) WITH { name: "Paula" } IN users
```

Target a different collection
-----------------------------

> The documents an `UPDATE` operations modifies can be in a different collection

Reviewer: ```suggestion
The documents an `UPDATE` operation modifies can be in a different collection
```

```aql
FOR i IN 1..1000
  UPDATE CONCAT('test', i)
  WITH { foobar: true } IN users
  OPTIONS { ignoreErrors: true }
```

> An update operation only update the attributes specified in `document` and
> leave other attributes untouched. Internal attributes (such as `_id`, `_key`, `_rev`,

Reviewer: ```suggestion
An update operation only updates the attributes specified in `document` and
leaves other attributes untouched. Internal attributes (such as `_id`, `_key`, `_rev`,
```

`_from` and `_to`) cannot be modified and are ignored when specified in `document`.
Updating a document modifies the document's revision number with a server-generated value.

### `keepNull`

> When updating an attribute to a `null` value, ArangoDB does not remove the attribute 

Reviewer: ```suggestion
When updating an attribute to the `null` value, ArangoDB does not remove the attribute 
```

from the document but stores this `null` value. To remove attributes in an update

> operation, set them to `null` and set the `keepNull` option to `false`:

Author: ```suggestion
operation, set them to `null` and set the `keepNull` option to `false`. This removes
the attributes you specify but not any previously stored attributes with the `null` value:
```

> 

Reviewer: It is not clear from the text whether in this case also attributes that have been set to `null` _previously_ are deleted.

Author: Added "This removes the attributes you specify but not any previously stored attributes with the `null` value"

```aql
FOR u IN users
  UPDATE u WITH { foobar: true, notNeeded: null } IN users
  OPTIONS { keepNull: false }
```

> The above query removes the `notNeeded` attribute from the documents and update

Reviewer: ```suggestion
The above query removes the `notNeeded` attribute from the documents and updates
```

the `foobar` attribute normally.

### `mergeObjects`

The option `mergeObjects` controls whether object contents are

> merged if an object attribute is present in both, the `UPDATE` query and in the 

Reviewer: ```suggestion
merged if an object attribute is present in both the `UPDATE` query and in the 
```

Reviewer: no comma

The `OLD` pseudo-value refers to the document revisions before the update, and `NEW` 
refers to document revisions after the update.

> Both ,`OLD` and `NEW`, contain all document attributes, even those not specified 

Reviewer: ```suggestion
Both `OLD` and `NEW` contain all document attributes, even those not specified 
```
